### PR TITLE
feat: add on_predicate decorator to handle 429

### PR DESCRIPTION
**Pull request by AlejandroUPC**
## Description

This MR aims to cover the case HTTP `429`  "Too Many Requests" ([RFC685](https://www.rfc-editor.org/rfc/rfc6585#section-4))i n a more specific way rather than the current way to handle any exception between the HTTP status code 400 and 500. This is motivated by the issue opened by @kgpayne [here](https://github.com/MeltanoLabs/tap-gitlab/issues/94).

## Implementation

The first decorator to be executed and added on top will be only evaluated when the HTTP code is 429 and then backoff whatever time is in the `Retry-After` header that gitlab returns (see [here](https://docs.gitlab.com/ee/administration/settings/user_and_ip_rate_limits.html#response-headers)).
The example is literally an implementation of the example supplied in the README.md from the backoff library [here](https://github.com/litl/backoff?tab=readme-ov-file#backoffruntime) + adding the same amount of max_times.

### CODE REVIEW

File: tap_gitlab/__init__.py [hunk 1]
```diff
@@ -248,11 +248,15 @@ def get_start(entity):
         STATE[entity] = CONFIG['start_date']
     return STATE[entity]
 
-
+@backoff.on_predicate(backoff.runtime,
+                      predicate=lambda r: r.status_code == 429,
+                      max_tries = 5, 
+                      vale=lambda r: int(r.headers.get("Retry-After")),
```
Comment (edgarrmondragon) [inline tap_gitlab/__init__.py]: is `vale` a typo?

Comment (AlejandroUPC) [reply tap_gitlab/__init__.py]: Good catch thanks

File: tap_gitlab/__init__.py [hunk 2]
```diff
@@ -248,11 +248,15 @@ def get_start(entity):
         STATE[entity] = CONFIG['start_date']
     return STATE[entity]
 
-
+@backoff.on_predicate(backoff.runtime,
+                      predicate=lambda r: r.status_code == 429,
+                      max_tries = 5,
```
Comment (edgarrmondragon) [inline tap_gitlab/__init__.py]: let's keep the styling as consistent as possible

```suggestion
                      max_tries=5, 
```

**Review by edgarrmondragon — APPROVED** [review]

**Comment by edgarrmondragon** [comment]
Thanks @AlejandroUPC!

**Pull request merged**